Explanations.txt

```diff
@@ -0,0 +1,92 @@
+7. Timer - is present in most of the versions of the game i have tried.
+
+Also need to add Unit and Property-based tests unique to the Haskell implementaton of the game.
+
+------------------------------------------------------------------------------------------
+
+First, I setup a project. I decided to use stack, as we have been primarily using it for Homework Assignments.
+Named the project Minesweeper. From the beginning, to follow modular programming and seperation of concerns 
+principles I have decided to Each module is designed to handle a specific part of the game’s functionality, 
+ensuring high cohesion and low coupling. This approach makes the code easier to understand, test, and maintain.
+For example, the Grid module focuses solely on grid-related operations. Moved Main.hs into app folder to seperate
+the entry point of the program from the library code (src/...)
+
+------------------------------------------------------------------------------------------
+
+Second, I started working on the Game Grid. First I made a new Algebraic Data Type (ADT) 
+for cells. It has only one constructor with multiple fields. Each field corresponds to
+possible contents of the cell. A cell can either have a revealed mine, a revealed cell
+with a number of adjacent mines, a flagged cell, or an unrevealed cell.
+
+----------------------------------------------------------------------
+Next worked on random generation of grids. Explanations behind certain choices made are in the comments
+
+randomUniqueIndices :: Int -> Int -> IO [Int]
+randomUniqueIndices n maxIndex = do
+  -- replicateM to repeat a monadic action multiple times and collect results in a list
+  indices <- replicateM n (randomRIO (0, maxIndex - 1))
+  -- nub to remove duplicating indices
+  return $ take n (nub indices)
+
+wanted to do this, but then if indices are duplicated and removed due to nub there would be less than n mines
+Instead used Sets since they already handle the duplicates.
+
+-------------------------------------------------------------------------------------------------
+
+Third, Revealing cells.
+
+When designing the revealCell function, I focused on implementing the cascading reveal,
+ where empty cells with no adjacent mines recursively reveal their neighbors until all connected 
+ empty regions are uncovered. I chose recursion for this because it naturally reflects the depth-first 
+ behavior of revealing cells in Minesweeper. Using foldl, I applied the revealCell function to all valid neighbors, 
+ progressively updating the grid without mutating state, making the cascading process both intuitive and efficient.
+
+----------------------------------------------------------------------------------------------------
+First playable version ready. 
+
+After playing for a bit found a lot of issues.
+One of them was that I forgot to take into account that the first move that the player makes should never turn out to be a mine.
```

Comment:
Well... The version of a minesweeper game I had in my childhood allowed the user to blow up at the first move.